### PR TITLE
[refactor](close)Full refactor async writer

### PR DESCRIPTION
**Pull request by yiguolei**
## Proposed changes

Issue Number: close #xxx

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by yiguolei** [comment]
run buildall

**Comment by yiguolei** [comment]
run buildall

**Comment by yiguolei** [comment]
run buildall

**Comment by yiguolei** [comment]
run p0

**Comment by yiguolei** [comment]
run compile

**Comment by yiguolei** [comment]
run buildall

**Comment by yiguolei** [comment]
run buildall

**Comment by yiguolei** [comment]
run buildall

File: be/src/pipeline/pipeline_x/operator.cpp [hunk 1]
```diff
@@ -541,20 +541,23 @@ Status AsyncWriterSink<Writer, Parent>::close(RuntimeState* state, Status exec_s
...
+        if (exec_status.ok()) {
+            _writer->force_close(state->is_cancelled() ? Status::Cancelled("Cancelled")
+                                                       : Status::Cancelled("force close"));
+        } else {
+            _writer->force_close(exec_status);
+        }
+        // If there is an error in process_block thread, then we should get the writer
+        // status before call force_close. For example, the thread may failed in commit
+        // transaction.
+        if (!st.ok()) {
```
Comment (Gabriel39) [inline be/src/pipeline/pipeline_x/operator.cpp:554]: RETURN_IF_ERROR

**Review by Gabriel39 — DISMISSED** [review]
LGTM

File: be/src/vec/sink/writer/async_result_writer.cpp [hunk 2]
```diff
@@ -87,8 +85,20 @@ std::unique_ptr<Block> AsyncResultWriter::_get_block_from_queue() {
 }
 
 void AsyncResultWriter::start_writer(RuntimeState* state, RuntimeProfile* profile) {
+    // Should set to false here, to
+    _writer_thread_closed = false;
+    // This is a async thread, should lock the task ctx, to make sure runtimestate and profile
+    // not constructed before the thread exit.
```
Comment (HappenLee) [inline be/src/vec/sink/writer/async_result_writer.cpp:91]: destructed ？or constructed？

**Comment by yiguolei** [comment]
run buildall

File: be/src/vec/sink/async_writer_sink.h [hunk 3]
```diff
@@ -105,21 +105,24 @@ class AsyncWriterSink : public DataSink {
         if (_writer) {
             // For pipeline engine, the writer is always closed in async thread process_block
             if (state->enable_pipeline_exec()) {
-                RETURN_IF_ERROR(_writer->get_writer_status());
+                Status st = _writer->get_writer_status();
+                if (exec_status.ok()) {
+                    _writer->force_close(state->is_cancelled() ? Status::Cancelled("Cancelled")
```
Comment (HappenLee) [inline be/src/vec/sink/async_writer_sink.h:110]: which case the cancel the writer for `force close`

**Review by xinyiZzz — APPROVED** [review]
LGTM

**Review by wangbo — APPROVED** [review]
LGTM

**Review by HappenLee — APPROVED** [review]
LGTM

**Pull request merged**